node/src/Pos/Recovery/Instance.hs

```diff
@@ -1,19 +1,49 @@
+-- | This module defines an instance of 'MonadRecoveryInfo'.
+
 module Pos.Recovery.Instance
        (
        ) where
 
 import           Universum
 
 import qualified Control.Concurrent.STM as STM
-import           Ether.Internal         (HasLens (..))
+import           Control.Monad.Except   (runExceptT, throwError)
 
 import           Pos.Context.Context    (RecoveryHeader, RecoveryHeaderTag)
-import           Pos.Recovery.Info      (MonadRecoveryInfo (..))
+import           Pos.Core               (HasCoreConstants, epochOrSlotG,
+                                         epochOrSlotToSlot, flattenSlotId)
+import           Pos.DB.Block           (MonadBlockDB)
+import           Pos.DB.DB              (getTipHeader)
+import           Pos.Recovery.Info      (MonadRecoveryInfo (..), SyncStatus (..))
+import           Pos.Slotting           (MonadSlots (getCurrentSlot))
+import           Pos.Util.Util          (HasLens (..))
 
-instance
-    ( Monad m, MonadIO m, MonadReader ctx m
-    , HasLens RecoveryHeaderTag ctx (RecoveryHeader ssc)
-    ) => MonadRecoveryInfo m where
-    recoveryInProgress = do
-        var <- view (lensOf @RecoveryHeaderTag)
-        isJust <$> atomically (STM.tryReadTMVar var)
+instance ( Monad m
+         , MonadIO m
+         , MonadBlockDB ssc m
+         , MonadSlots ctx m
+         , MonadReader ctx m
+         , HasLens RecoveryHeaderTag ctx (RecoveryHeader ssc)
+         , HasCoreConstants
+         ) =>
+         MonadRecoveryInfo m where
+    getSyncStatus lagBehindParam =
+        fmap convertRes . runExceptT $ do
+            recoveryInProgress >>= \case
+                False -> pass
+                True -> throwError SSDoingRecovery
+            curSlot <- note SSUnknownSlot =<< getCurrentSlot
+            tipHeader <- getTipHeader @ssc
+            let tipSlot = epochOrSlotToSlot (tipHeader ^. epochOrSlotG)
+            let slotDiff = flattenSlotId curSlot - flattenSlotId tipSlot
+            unless (slotDiff < fromIntegral lagBehindParam) $
+                throwError
+                    SSLagBehind
+                    {sslbCurrentSlot = curSlot, sslbTipSlot = tipSlot}
+      where
+        recoveryInProgress = do
```

Comment:
can you rename it? it's quite confusing, I confused it w/ global `recoveryInProgress` and was :cho:

Reply:
Ok.

Reply:
>Ok.
>merged

:peka:

Reply:
I wrote this:
> I am gonna merge it now, because CI takes a lot of time to pass nowadays. There will be another PR to please @pva701.

Reply:
sorry, didn't notice